### PR TITLE
Add sha256t_hash_newtype macro to create tagged hash newtypes

### PR DESCRIPTION
**Pull request by stevenroose**

### CODE REVIEW

File: src/sha256t.rs [hunk 1]
```diff
@@ -78,6 +78,28 @@ impl<T: Tag> HashTrait for Hash<T> {
...
 
+/// Macro used to define a newtype tagged hash.
+/// It creates two public types:
+/// - a sha246t::Tag struct,
+/// - a sha256t::Hash type alias.
+#[macro_export]
+macro_rules! sha256t_hash_newtype {
+    ($newtype:ident, $tag:ident, $midstate:ident, $midstate_len:expr, $docs:meta, $reverse:expr) => {
+        /// The tag used for [$newtype].
+        #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Default, Hash)]
```
Comment (dr-orlovsky) [inline src/sha256t.rs:89]: What about `Debug`?

**Review by apoelstra — APPROVED** [review]
utACK

File: src/sha256t.rs [hunk 2]
```diff
@@ -78,6 +78,28 @@ impl<T: Tag> HashTrait for Hash<T> {
     }
 }
 
+/// Macro used to define a newtype tagged hash.
+/// It creates two public types:
+/// - a sha246t::Tag struct,
+/// - a sha256t::Hash type alias.
+#[macro_export]
+macro_rules! sha256t_hash_newtype {
```
Comment (dr-orlovsky) [inline src/sha256t.rs:86]: May be `sha2561_newtype`, since `sha` already implies that it is a hash?

Comment (apoelstra) [reply src/sha256t.rs:86]: I think the name is for symmetry with `hash_newtype`

File: src/sha256t.rs (hunk 1 above)
Comment (stevenroose) [reply src/sha256t.rs:89]: added

File: src/sha256t.rs (hunk 2 above)
Comment (stevenroose) [reply src/sha256t.rs:86]: could be a tag... :) idk I think it's good to be inline with `hash_newtype`?

**Review by dr-orlovsky — APPROVED** [review]

**Review by apoelstra — APPROVED** [review]
re-utACK

**Comment by dr-orlovsky** [comment]
confirm my ACK

**Comment by dr-orlovsky** [comment]
Just understood that this macro uses `hash_newtype!`, which implements `Display` and `FromStr` defaulting to reading/creating hexadecimal representation. However for certain tagged hashes types I use Bech32 encoding, and it's impossible to change that with this macro...

**Comment by stevenroose** [comment]
Why do you use bech32? Hmm yeah well the reason I had these in rust-bitcoin in a taproot module was partly to make sure this wouldn't grow larger than just taproot usage. But still I think as long as there is no Bitcoin usage of tagged hashes that serializes differently, I think it's fine here.

It's really not super hard to specify an alternative macro in your project to create bech32 tagged hash newtypes, right?

**Pull request merged**